### PR TITLE
Initial design specification for adding pods to a specific subnet/VPC

### PR DESCRIPTION
**Pull request by phudtran**
This PR specifies the interface for users to assign pods to specific VPCs and Subnets.

### CODE REVIEW

File: docs/design/mp_pod_to_vpc_subnet.md [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+      - containerPort: 443
+```
+
+#### Requirements
+- The Parent VPC must exist.
+- The VPC must have an existing subnet.
+
+The pod is invalid if the VPC does not exist.
+
+### Placing a pod into a specific VPC and Subnet
```
Comment (pdgetrf) [inline docs/design/mp_pod_to_vpc_subnet.md:104]: this may be beyond the scope of this design, but for workloads like deployment or job, user should also be able to specify vpc and subnet, and that could lead to more scenarios:

1. all pods from the same deployment belong to the same subnet
2. all pods from the same deployment belong to different subnets

File: docs/design/mp_pod_to_vpc_subnet.md [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+Users can place a pod into a specific VPC and subnet by specifying the required VPC and subnet annotation like below.
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: nginx
+  annotations:
+    mizar.com: {"vpc": "vpc0", "subnet": "subnet0"}
```
Comment (pdgetrf) [inline docs/design/mp_pod_to_vpc_subnet.md]: Inside XDP programs (for example, the transit XDP), I think VPCs and subnets are referenced/identified by VNI instead of names. In the  In the edge cluster situation, there will be cases where two subnets from two different clusters  belong to the same VPC. When a packet travels from one cluster to the other, VNI will be used in the process for matching, and if the VPC on one cluster has a different VNI from the VPC on the other cluster, it could drop the packet even though the VPCs on both clusters have the same name. In fact, there could also be a case where two VPCs in two cluster have the same name but with two different VNIs. In this case they are different VPCs. 

In summary, the VPC name -> VNI mapping works correctly in one cluster, but cross-cluster we will need to answer what is the unique identifier for a VPC? For example, if two VPC in different clusters have the same name but different VNI, is that considered the same VPC? What about different VPC names but the same VNI?

This needs to be answered or it will lead to complication down the road. I don't have an answer yet and we can discuss.

Comment (pdgetrf) [reply docs/design/mp_pod_to_vpc_subnet.md]: I'd tent to think VPC name is the unique identifier cross-clusters. The different VNI could be converted by the edge gateway.

**Review by pdgetrf — COMMENTED** [review]
left 2 questions.

File: docs/design/mp_pod_to_vpc_subnet.md [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+
+An example is given below.
+
+```yaml
+apiVersion: "mizar.com/v1"
+kind: Vpc
+metadata:
+  name: vpc0
+spec:
+  ip: "12.0.0.0"
```
Comment (jshaofuturewei) [inline docs/design/mp_pod_to_vpc_subnet.md:27]: I found there is no vni in subnet and vpc.  Do we still need vni or vni is not user defined any more?

File: docs/design/mp_pod_to_vpc_subnet.md [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+- The Parent VPC must exist.
+- The new subnet must not yet already exist within the VPC.
+
+## Pod to VPC and Subnet designation
+
+Three options are defined for assigning pods to specific VPCs.
+
+### Placing a pod into the default VPC
+
+Creating a standard pod with no Mizar specific annotation will assign it to the default VPC that is created at Mizar setup.
```
Comment (jshaofuturewei) [inline docs/design/mp_pod_to_vpc_subnet.md]: default subnet as well?

**Comment by chenqianfzh** [comment]
> This PR specifies the interface for users to assign pods to specific VPCs and Subnets

Is it supported to assign an existing pod to a VPC/Subnet? Or update the VPC/Subnet of a running pod? I guess it is probably not for now. Could you please call it out in the design doc either it is supported or not?

File: docs/design/mp_pod_to_vpc_subnet.md (hunk 3 above)
Comment (phudtran) [reply docs/design/mp_pod_to_vpc_subnet.md:27]: VNI is no longer user defined.

File: docs/design/mp_pod_to_vpc_subnet.md (hunk 4 above)
Comment (phudtran) [reply docs/design/mp_pod_to_vpc_subnet.md]: Yes! Thanks for the catch. Will update

**Comment by phudtran** [comment]
> > This PR specifies the interface for users to assign pods to specific VPCs and Subnets
> 
> Is it supported to assign an existing pod to a VPC/Subnet? Or update the VPC/Subnet of a running pod? I guess it is probably not for now. Could you please call it out in the design doc either it is supported or not?

That's right this is to supported for now. I will add this to the doc.

File: docs/design/mp_pod_to_vpc_subnet.md [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+  dividers: 2
+```
+
+#### Requirements
+- The CIDR range is valid.
+- The VPC must not already exist.
+
+## Custom Subnet Creation
+
+When creating a subnet users must define the subnet's name, IP, prefix  the number of bouncers, and the VPC that it belongs to.
```
Comment (vinaykul) [inline docs/design/mp_pod_to_vpc_subnet.md:38]: Can we consider supporting default values for number of dividers and bouncers? (Default 1, if not specified). Effectively, it would make these fields optional.

File: docs/design/mp_pod_to_vpc_subnet.md [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+
+Users can place a pod into a specific VPC by specifying the VPC annotation like below.
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: nginx
+  annotations:
+    mizar.com: {"vpc": "vpc0"}
```
Comment (vinaykul) [inline docs/design/mp_pod_to_vpc_subnet.md]: I'm not sure if this works. Annotations are usually specified as key: value.
e.g. mizar.com/vpc: vpc0

File: docs/design/mp_pod_to_vpc_subnet.md [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+    image: nginx
+    ports:
+      - containerPort: 443
+```
+
+#### Requirements
+- The Parent VPC must exist.
+- The VPC must have an existing subnet.
+
+The pod is invalid if the VPC does not exist.
```
Comment (vinaykul) [inline docs/design/mp_pod_to_vpc_subnet.md:102]: I'm not sure if it is possible to define a user-visible behavior for the pod state if this occurs. Can the operator update the pod status or write to the event stream?

File: docs/design/mp_pod_to_vpc_subnet.md (hunk 2 above)
Comment (vinaykul) [inline docs/design/mp_pod_to_vpc_subnet.md]: This would more likely be:
```golang
  annotations:
    mizar.com/vpc: vpc0
    mizar.com/subnet: subnet0
```

Please double-check if the above format is also supported for annotations.

File: docs/design/mp_pod_to_vpc_subnet.md (hunk 7 above)
Comment (jshaofuturewei) [reply docs/design/mp_pod_to_vpc_subnet.md:102]: I am thinking if we create pods with any existing vpcs or subnets and then create those vpcs and subnets, will we update those running pods into invalid states or just leave them alone?

File: docs/design/mp_pod_to_vpc_subnet.md (hunk 5 above)
Comment (phudtran) [reply docs/design/mp_pod_to_vpc_subnet.md:38]: Sure I can add that.

File: docs/design/mp_pod_to_vpc_subnet.md (hunk 6 above)
Comment (phudtran) [reply docs/design/mp_pod_to_vpc_subnet.md]: You're right will fix.

File: docs/design/mp_pod_to_vpc_subnet.md (hunk 2 above)
Comment (phudtran) [reply docs/design/mp_pod_to_vpc_subnet.md]: Agreed will fix.

File: docs/design/mp_pod_to_vpc_subnet.md (hunk 7 above)
Comment (phudtran) [reply docs/design/mp_pod_to_vpc_subnet.md:102]: Usually we retry a number of times before giving up in the operator. 
It would be preferable to just return in error to the user immediately, similar to syntax errors in a YAML file but I'm not sure if it's possible.

File: docs/design/mp_pod_to_vpc_subnet.md (hunk 1 above)
Comment (phudtran) [reply docs/design/mp_pod_to_vpc_subnet.md:104]: Agreed, I believe we can tackle this the same way as we do with pods.

If no VPC specified -> default VPC and default subnet for all pods.
If VPC specified but no subnet specified -> default subnets for all pods.

File: docs/design/mp_pod_to_vpc_subnet.md (hunk 2 above)
Comment (pdgetrf) [reply docs/design/mp_pod_to_vpc_subnet.md]: Discussed with @vinaykul  and @phudtran offline about this. The conclusion at the moment is to have the VPC name as the unique identifier of VPC. This means within a cluster, Mizar will generate VNI (VPC id) locally, and Edge Gateway will perform translation of VNI in and out of its cluster based on VPC name. VPC name needs to be global unique which will be enforced by the root level (cloud) cluster which all edge clusters connect with.

Comment (vinaykul) [reply docs/design/mp_pod_to_vpc_subnet.md]: Sounds good. We'll merge this doc and work on this design. If we need to change something, we can always revisit.

**Comment by vinaykul** [comment]
LGTM
